clusterloader2/cmd/clusterloader.go

```diff
@@ -307,20 +307,53 @@ func main() {
 
 	testReporter := test.CreateSimpleReporter(path.Join(clusterLoaderConfig.ReportDir, "junit.xml"), "ClusterLoaderV2")
 	testReporter.BeginTestSuite()
+
+	var contexts []test.Context
+	var testConfigs []*api.Config
+
 	if testSuiteConfigPath != "" {
 		testSuite, err := config.LoadTestSuite(testSuiteConfigPath)
 		if err != nil {
 			klog.Exitf("Error while reading test suite: %v", err)
 		}
+
 		for i := range testSuite {
 			clusterLoaderConfig.TestScenario = testSuite[i]
-			runSingleTest(f, prometheusFramework, testReporter)
+			ctx, errList := test.CreateTestContext(f, prometheusFramework, &clusterLoaderConfig, testReporter)
+			if !errList.IsEmpty() {
+				klog.Exitf("Test context creation failed: %s", errList.String())
+			}
+			contexts = append(contexts, ctx)
+
+			testConfig, errList := test.CompileTestConfig(ctx)
+			if !errList.IsEmpty() {
+				klog.Exitf("Test compliation failed: %s", errList.String())
+			}
+			testConfigs = append(testConfigs, testConfig)
+		}
+
+		for i := range testSuite {
+			runSingleTest(contexts[i], testConfigs[i])
 		}
 	} else {
 		for i := range testConfigPaths {
 			clusterLoaderConfig.TestScenario.ConfigPath = testConfigPaths[i]
 			clusterLoaderConfig.TestScenario.OverridePaths = testOverridePaths
```

Comment:
Hmm, I'm not sure whether changing the clusterLoaderConfig and then passing it by reference (pointer) is correct. I think it's not, as the second iteration will modify the context we created in the first iteration (because it'll modify the clusterLoaderConfig which is referenced in the testContext). Do I miss something? 

If I'm right, I believe we should get rid of the TestScenario field from the ClusterLoaderConfig and move it to the TestContext, thoughts?

Reply:
Good catch, I think you are right. I should be able to move it to test context (the alternative is to make copies of ClusterLoaderConfig which we are trying to avoid)

Reply:
I've decoupled TestScenario from ClusterLoaderConfig here https://github.com/kubernetes/perf-tests/pull/1680/commits/b793c9e39c2ca7433520d629aed578c06b3c1e3e

Reply:
Did a bit of digging to understand why it still worked for the previous [run](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/perf-tests/1680/pull-perf-tests-clusterloader2/1354805480483983360). It was because the testConfigs were [compiled](https://github.com/kubernetes/perf-tests/pull/1680/files#diff-d6057f57c97c93f50c6e310f19970261bbd5d145d2f9682473e3749afe7abb97R349) and passed directly to the test executor. So leaving TestScenario in ClusterLoaderConfig indeed doesn't make sense any more.